File: library/src/main/java/com/google/maps/android/data/kml/KmlUtil.java
package com.google.maps.android.data.kml;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Utility methods for KML
 */
public class KmlUtil {

    /**
     * Substitute property values in BalloonStyle text template
     *
     * @param template text template
     * @param placemark placemark to get property values from
     * @return string with property values substituted
     */
    public static String substituteProperties(String template, KmlPlacemark placemark) {
        StringBuffer sb = new StringBuffer();
        Pattern pattern = Pattern.compile("\\$\\[(.+?)]");
        Matcher matcher = pattern.matcher(template);
        while (matcher.find()) {
            String property = matcher.group(1);
            String value = placemark.getProperty(property);
            if (value != null) {
                matcher.appendReplacement(sb, value);
            }
        }
        matcher.appendTail(sb);
        return sb.toString();
    }
}

File: data/src/main/java/com/google/maps/android/data/parser/kml/KmlUtil.kt
package com.google.maps.android.data.parser.kml

/**
 * Utility functions for parsing KML files.
 */
internal object KmlUtil {
    /**
     * Coerces a string value to a boolean, accepting "true", "false", "1", and "0".
     *
     * @param value The string to coerce.
     * @return `true` if the string is "true" or "1", `false` otherwise.
     */
    fun toBoolean(value: String?): Boolean {
        return value != null && (value.equals("true", ignoreCase = true) || value == "1")
    }
}